### PR TITLE
clipboard,win32: rewrite clipboard functions to handle errors

### PR DESCRIPTION
**Pull request by vtjnash**
Per comment (although it failed to note that we were previously always leaking the memory).

### CODE REVIEW

**Review by musm — APPROVED** [review]

File: stdlib/InteractiveUtils/src/clipboard.jl [hunk 1]
```diff
@@ -26,89 +27,113 @@ if Sys.isapple()
...
         end
-        Base.windowserror(:OpenClipboard, 0==ccall((:OpenClipboard, "user32"), stdcall, Cint, (Ptr{Cvoid},), C_NULL))
-        Base.windowserror(:EmptyClipboard, 0==ccall((:EmptyClipboard, "user32"), stdcall, Cint, ()))
         x_u16 = Base.cwstring(x)
+        pdata = Ptr{UInt16}(C_NULL)
+        function cleanup(cause)
+            errno = cause == :success ? UInt32(0) : Libc.GetLastError()
+            if cause !== :OpenClipboard
+                if cause !== :success && pdata != C_NULL
+                    ccall((:GlobalFree, "kernel32"), stdcall, Cint, (Ptr{UInt16},), pdata)
```
Comment (musm) [inline stdlib/InteractiveUtils/src/clipboard.jl]: check return status  ?

File: stdlib/InteractiveUtils/src/clipboard.jl [hunk 2]
```diff
@@ -26,89 +27,113 @@ if Sys.isapple()
...
-    # TODO: these functions leak memory and memory locks if they throw an error
     function clipboard(x::AbstractString)
         if Base.containsnul(x)
             throw(ArgumentError("Windows clipboard strings cannot contain NUL character"))
         end
-        Base.windowserror(:OpenClipboard, 0==ccall((:OpenClipboard, "user32"), stdcall, Cint, (Ptr{Cvoid},), C_NULL))
-        Base.windowserror(:EmptyClipboard, 0==ccall((:EmptyClipboard, "user32"), stdcall, Cint, ()))
         x_u16 = Base.cwstring(x)
+        pdata = Ptr{UInt16}(C_NULL)
+        function cleanup(cause)
```
Comment (musm) [inline stdlib/InteractiveUtils/src/clipboard.jl]: I suppose it's safe to call GlobalFree multiple names on pdata ?

File: stdlib/InteractiveUtils/src/clipboard.jl [hunk 3]
```diff
@@ -26,89 +27,113 @@ if Sys.isapple()
...
+    const _clipboard_copy = Dict(
             :xsel  => Sys.islinux() ?
-                `xsel --nodetach --input --clipboard` : `xsel -c`,
+                `xsel --nodetach --input --clipboard` :
+                `xsel -c`,
             :xclip => `xclip -silent -in -selection clipboard`,
-        ),
-        :paste => Dict(
+        )
+    const _clipboard_paste = Dict(
```
Comment (musm) [inline stdlib/InteractiveUtils/src/clipboard.jl]: I suppose this was done due to performance reasons

File: stdlib/InteractiveUtils/src/clipboard.jl (hunk 2 above)
Comment (vtjnash) [reply stdlib/InteractiveUtils/src/clipboard.jl]: Absolutely not. We should only call this exactly once.

File: stdlib/InteractiveUtils/src/clipboard.jl (hunk 1 above)
Comment (vtjnash) [reply stdlib/InteractiveUtils/src/clipboard.jl]: it's fine, it won't fail

**Pull request merged**